src/AddNewComponent.tsx

```diff
@@ -0,0 +1,113 @@
+import { useEffect, useRef, useState } from "preact/hooks";
+import PlusIcon from "./icons/PlusIcon.tsx";
+import Button from "./ui/Button.tsx";
+import Modal from "./ui/Modal.tsx";
+import ComponentPreviewList from "./ComponentPreviewList.tsx";
+import type { ComponentPreview } from "../editor.tsx";
+
+function ToCItem({ component, componentLabel }: ComponentPreview) {
+  return (
+    <li key={component}>
+      <a
+        class="font-medium  text-gray-500 hover:text-black"
+        href={`#${component}`}
+      >
+        {componentLabel}
+      </a>
+    </li>
+  );
+}
+
+interface Props {
+  onAddComponent: (componentName: string) => void;
+}
+
+export default function AddNewComponent({ onAddComponent }: Props) {
+  const targetRef = useRef<HTMLDivElement>(null);
+  const [openModal, setOpenModal] = useState<boolean>(false);
+  const [tocComponents, setToCComponents] = useState<
+    ComponentPreview[]
+  >([]);
+
+  useEffect(function initObserverIframes() {
+    if (!targetRef.current || !(openModal && tocComponents.length > 0)) {
+      return;
+    }
+
+    const target = targetRef.current;
+
+    const observer = new IntersectionObserver((entries) => {
+      entries.forEach((entry) => {
+        if (entry.intersectionRatio > 0) {
+          const iframe = (entry.target as HTMLIFrameElement);
+
+          // Prevent set the src and make the browser fetch the page multiple times
+          if (!iframe.src) {
+            iframe.src = iframe.dataset.src ?? "";
+          }
+        }
+      });
+    }, {
+      root: target,
+    });
+
+    const iframes = target.getElementsByTagName("iframe");
+    for (const iframe of iframes) {
+      observer.observe(iframe);
+    }
+
+    return () => {
+      observer.disconnect();
+    };
+    // tocComponents are the components fetched by the ComponentsPreviewList and they doesn't exist when openModal = true at the first time
```

Comment:
I find it a bit hard to understand what this does. I believe it's a lazy-loading mechanism for a list of iframes that is scrollable.

Might be a good idea to comment on the function or make the function name more describing.

Reply:
Yeah, totally agree. I did this to avoid fetching multiple pages, but now the pages are cached and immutable. So, this isn't a problem.